eth/handler.go

```diff
@@ -716,6 +722,64 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 			}
 		}
 
+	case msg.Code == NewTxHashesMsg && p.version >= eth65:
+		// New transaction announcement arrived, make sure we have
+		// a valid and fresh chain to handle them
+		if atomic.LoadUint32(&pm.acceptTxs) == 0 {
+			break
+		}
+		var hashes []common.Hash
+		if err := msg.Decode(&hashes); err != nil {
+			return errResp(ErrDecode, "msg %v: %v", msg, err)
+		}
+		// Schedule all the unknown hashes for retrieval
+		for _, hash := range hashes {
+			// Mark the hashes as present at the remote node
+			p.MarkTransaction(hash)
+
+			// Filter duplicated transaction announcement.
+			if pm.hasTx(hash) {
+				continue
+			}
+			pm.txFetcher.Notify(p.id, hash, time.Now(), p.RequestTxs)
+		}
+	case msg.Code == GetTxsMsg && p.version >= eth65:
+		// Decode the retrieval message
+		msgStream := rlp.NewStream(msg.Payload, uint64(msg.Size))
+		if _, err := msgStream.List(); err != nil {
+			return err
+		}
+		// Gather state data until the fetch or network limits is reached
+		var (
+			hash  common.Hash
+			bytes int
+			txs   []rlp.RawValue
+		)
+		for bytes < softResponseLimit && len(txs) < fetcher.MaxTransactionFetch {
```

Comment:
Retrospectively, I'm not sure it's a good idea to use fixed number limits (this prevents us in the downloader to play with different request sizes). Rather than enforcing a global messages size limit `bytes < softResponseLimit` and a local item count limit `len(txs) < fetcher.MaxTransactionFetch`, we should imho have specific limits on data types. E.g. we could set a `softTransactionResponseLimit` to 128KB and only enforce that. It's small enough that you can't go wild with thousands of lookups, but more flexible than limiting to 64 txs only (e.g. if I only do transfers, might as well send back more than 64).

As for fetcher.MaxTransactionFetch, we can still keep that, it's fine to limit on the requester side, but separating the client and server limits allows us to play with them in the future.

Reply:
Hmm, but we need to avoid remote nodes requesting inifnite none existing txs, forcing DB reads. Though the original code doesn't handle that either.